devdoc/channel_2.md

```diff
@@ -0,0 +1,72 @@
+# channel_2 requirements:
+
+## Overview
+
+The current implementation of `channel` has certain pitfalls. `channel_2` aims to address these pitfalls by making the following changes:
+
+1. Introduce a `channel_2_reset` API: This API will allow the user to reset the `channel_2` to a clean state. It will cause all pending operations to be abandoned.
+
+2. Remove `ASYNC_OP`: The current implementation of `channel` provides the user with an `ASYNC_OP` when an operation (`channel_push` or `channel_pull`) is initiated. The user can call `async_op_cancel` to do a best-effort cancellation of the operation. There are 2 issues with this:
+    - Complicates object lifetimes: The lifetime of the operation must be linked to the `ASYNC_OP`and the `channel`, since each can be independently destroyed by the user.
+    - Inconsistent callback order: Consider when there are 3 `PUSH` operations in the `channel`: [`op1`, `op2`, `op3`] with data [`data1`, `data2`, `data3`] respectively.
+    
+    The receiver calls `channel_pull` 3 times and the sender cancels `op2` concurrently.
+    
+    The callback for `op1` is called with `data1` and `CHANNEL_CALLBACK_RESULT_OK`. 
+
+    The callback for `op2` is called with `NULL`  and `CHANNEL_CALLBACK_RESULT_CANCELLED`.
+
+    The callback for `op3` is called with `data3` and `CHANNEL_CALLBACK_RESULT_OK`.
+
+    The receiver will now contain [`data1`, `data3`].
+
+    This order of the data on the receiver side of the `channel` is not the same as the order of the `PUSH` operations on the sender side. This is not desirable.
+
+    - Insufficient value: There is also not much value in providing the user with an `ASYNC_OP` for a single operation. We have not found a use case where the user would need to cancel a single operation while maintaining the other operations active. The user can always abandon all operations by calling `channel_2_reset`.
```

Comment:
>We have not found a use case where the user would need to cancel a single operation while maintaining the other operations active. [](http://example.com/codeflow?start=120&length=130)

I don't think that cancelling operations is bad generally.
This being an attempt to a rather generic module, it sounds to me that the intent to have these operations nice and cancellable is commendable.

I can imagine a scenario where the waiters we use for read and wait would use channel for example.
So this flip flop between let's have cancellation or let's not have cancellation for individual operations seems a bit premature.

I think the right way would be to do the following:
- when the channel is closed, cancel all outstanding operations.
- switch state to closed.

Notice I am not using reset, as I think of reset being sugar for (close+open).